File: 백준/Gold/28132. 기벡을 안배운다고？/기벡을 안배운다고？.cc
#include <bits/stdc++.h>

using namespace std;

int gcd(int a, int b)
{
    return b == 0 ? a : gcd(b, a % b);
}

int main(void)
{
    ios::sync_with_stdio(0);
    cin.tie(0);

    int N;
    cin >> N;

    int cnt = 0;
    long long sum = 0;

    map<pair<int, int>, int> mp;

    for (int i = 0; i < N; i++)
    {
        int x, y;
        int tmpx, tmpy;
        cin >> x >> y;

        // 영벡터
        if (x == 0 && y == 0)
        {
            cnt++;
            continue;
        }
        // 단위벡터
        else if (x == 0)
        {
            tmpx = 0;
            tmpy = 1;
        }
        else if (y == 0)
        {
            tmpx = 1;
            tmpy = 0;
        }
        // 나머지 벡터
        else
        {
            tmpx = x / gcd(x, y);
            tmpy = y / gcd(x, y);
        }

        sum += mp[{tmpx, tmpy}];
        mp[{-tmpy, tmpx}]++;
        mp[{tmpy, -tmpx}]++;
    }

    cout << 1LL * sum + 1LL * cnt * (cnt - 1) / 2 + 1LL * cnt * (N - cnt);

    return 0;
}

File: 백준/Gold/28132. 기벡을 안배운다고？/기벡을 안배운다고？.py
import sys
from math import gcd
from collections import defaultdict

input = sys.stdin.read

def main():
    data = input().splitlines()
    N = int(data[0])

    cnt = 0
    sum = 0
    ump = defaultdict(int)

    for i in range(1, N + 1):
        x, y = map(int, data[i].split())

        # 영벡터
        if x == 0 and y == 0:
            cnt += 1
            continue
        # 단위벡터
        elif x == 0:
            tmpx, tmpy = 0, 1
        elif y == 0:
            tmpx, tmpy = 1, 0
        # 나머지
        else:
            g = gcd(x, y)
            tmpx, tmpy = x // g, y // g

        sum += ump[(tmpx, tmpy)]
	# 90도 벡터 추가
        ump[(-tmpy, tmpx)] += 1
        ump[(tmpy, -tmpx)] += 1

    print(sum + cnt * (cnt - 1) // 2 + cnt * (N - cnt))

if __name__ == "__main__":
    main()